app/jobs/export_job.rb

```diff
@@ -5,34 +5,56 @@ class ExportJob < ApplicationJob
   queue_as :exports
   include ImportExport
 
-  # Limits number of Patient records to be considered for a single exported file to handle maximum file size limit.
-  # Adds additional files as needed if exceeds batch size.
-  OUTER_BATCH_SIZE = ENV['EXPORT_OUTER_BATCH_SIZE']&.to_i || 10_000
```

Comment:
so I've been thinking about this more and I'm wondering if it would be safer for us to keep some kind of limit here (maybe `100_000` or something) to prevent the memory usage from spiking out of control if a user exports several hundred thousands of records at a time. @holmesie thoughts?

Reply:
I can run a benchmark for this but just to illustrate what I think is going to happen with the memory usage is that the spikes seen in the image below are going to be much larger and the upward trend in the beginning is going to continue going up. This is caused by the fact that the entire file has to be loaded into memory on `read_string` when it's time to write it to the db (or somewhere else in this case)

![36k_custom](https://user-images.githubusercontent.com/9956561/113587558-f46ccb80-95fc-11eb-9228-1dda890536fa.png)

Reply:
Is there a way for us to stream the file up to S3?

Reply:
^ that would be great if possible because it would mean we don't even have to read the file into memory at all, resulting in no spike at all

Reply:
For CSV export the upper limit of memory usage is going to be the size of the IO itself. Excel upper limit could be larger because of the more complex Ruby objects. Calling `read_string` would increase memory size by the size of the Excel file. Activestorage upload chunks the IO for upload as well. I'm not sure what you're talking about from the link you posted. We are following the pattern in that section almost exactly.

Reply:
My main question here is that if `read_string` loads the entire file into memory or if we will still be streaming from disk. 

It looks like from the fast_excel source code that `read_string` will return a `File` object (https://github.com/Paxa/fast_excel/blob/master/lib/fast_excel.rb#L393) so I'm not convinced we need to wrap the `workbook.read_string` call in `String.IO(...)` on L106 of `import_export.rb`

Reply:
Cool let's test that out

Reply:
That line isn't returning a File it is returning the result of `read` so we still need `StringIO`

Reply:
so if that is returning the results of read does that mean the entire file will be loaded into memory?

Reply:
the entire file is being loaded into memory by calling `read_string`. if we could somehow upload the file directly from the fs to activestorage without opening the file, we could avoid that

Reply:
So I just finished doing some memory profiling for all the export types with 36k, 72k, and 144k monitorees and it seems like there's no upper limit to how much memory any export is going to take without any kind of outer batching.

I've been looking through new relic to see how long some of the exports take and I'm seeing some that are over 10 minutes which equates to over 300k monitorees or 8GB of memory used for a single export. It's also possible and likely that more monitorees will be exported in the future, making this somewhat of a concern to me.

@holmesie thoughts on putting the outer batching back in unless we can figure out some other way to keep the memory usage down?

![Screen Shot 2021-04-08 at 10 22 12 PM](https://user-images.githubusercontent.com/9956561/114119596-f4005900-98b8-11eb-9798-3e801088500f.png)

Reply:
If the `read_string` problem can't be avoided, yes - we would want to avoid that situation (perhaps only do so if some upper bound is hit). Let's discuss at standup.

Reply:
this issue is resolved by these changes: https://github.com/SaraAlert/SaraAlert/pull/550/commits/9b11d7efb97cffd72badf5088ef54854040aafde